### PR TITLE
Gateway Submit

### PR DESCRIPTION
**Pull request by CyonAlexRDX**
Fear not the LoC delta: added JSON vectors...

- Major file restructure of Gateway files
- Add JSON test vectors for GW models - with serde testing of them
- Implement new endpoint: `/transaction/submit`
- Implement method on GatewayClient:  `submit_notarized_transaction`
- More tests

### CODE REVIEW

File: apple/Sources/Sargon/Antennas/Network/NetworkAntenna+URLSession.swift [hunk 1]
```diff
@@ -28,16 +29,17 @@ extension NetworkResponse {
 }
 
 extension URLSession: NetworkAntenna {
-	
-	public func makeRequest(
+
+	public func executeNetworkRequest(
```
Comment (CyonAlexRDX) [inline apple/Sources/Sargon/Antennas/Network/NetworkAntenna+URLSession.swift:33]: I think `execute` is a better name than `make`

File: build.rs [hunk 2]
```diff
@@ -19,6 +19,16 @@ pub fn main() {
...
         fixtures_vector_path.display()
     );
+    let fixtures_models_path = fixtures_path.join("models");
+    println!(
+        "cargo:rustc-env=FIXTURES_MODELS={}/",
+        fixtures_models_path.display()
+    );
+    let fixtures_gw_models_path = fixtures_models_path.join("gateway");
+    println!(
+        "cargo:rustc-env=FIXTURES_MODELS_GW={}/",
```
Comment (CyonAlexRDX) [inline build.rs:29]: making it easier to unit test Gateway Models against newly added JSON files

File: fixtures/models/gateway/state/request_entity_details__single_account_no_assets.json
Comment (CyonAlexRDX) [inline fixtures/models/gateway/state/request_entity_details__single_account_no_assets.json:1]: no need to even look at all the JSON files 🤪

File: src/core/assert_json.rs [hunk 3]
```diff
@@ -1,8 +1,47 @@
...
 use assert_json_diff::assert_json_include;
 use core::fmt::Debug;
 use pretty_assertions::{assert_eq, assert_ne};
-use serde::{de::DeserializeOwned, ser::Serialize};
+use serde::de::DeserializeOwned;
 use serde_json::Value;
+use thiserror::Error as ThisError;
+
+#[derive(Debug, ThisError)]
+pub enum TestingError {
```
Comment (CyonAlexRDX) [inline src/core/assert_json.rs:11]: moved from `vector` integration tests, now used by that integration test but also by GW model unit tests :)

File: src/core/types/keys/private_key.rs [hunk 4]
```diff
@@ -44,13 +44,42 @@ impl PrivateKey {
...
-            PrivateKey::Ed25519(key) => PublicKey::Ed25519 {
-                value: key.public_key(),
+            PrivateKey::Ed25519(key) => PublicKey::Ed25519(key.public_key()),
+            PrivateKey::Secp256k1(key) => {
+                PublicKey::Secp256k1(key.public_key())
+            }
+        }
+    }
+
+    pub fn sign(&self, msg_hash: &Hash) -> Signature {
```
Comment (CyonAlexRDX) [inline src/core/types/keys/private_key.rs:54]: same method that existed on `Secp256k1PrivateKey` and `Ed25519PrivateKey` but on the general type `PrivateKey`.

File: src/core/types/keys/public_key.rs [hunk 5]
```diff
@@ -12,14 +12,15 @@ use crate::prelude::*;
     Hash,
     PartialOrd,
     Ord,
+    derive_more::Display,
     uniffi::Enum,
 )]
 pub enum PublicKey {
     /// An Ed25519 public key used to verify cryptographic signatures.
-    Ed25519 { value: Ed25519PublicKey },
+    Ed25519(Ed25519PublicKey),
```
Comment (CyonAlexRDX) [inline src/core/types/keys/public_key.rs:20]: not needed by UniFFI anymore 🥳

File: src/gateway_api/client/gateway_client.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
+use crate::prelude::*;
+
+/// An HTTP client for consuming the Radix ⛩️ Gateway API ([docs]).
```
Comment (CyonAlexRDX) [inline src/gateway_api/client/gateway_client.rs:3]: refactor - moved file and split out other types, and also added documentation.

File: src/gateway_api/client/gateway_client.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+#[derive(Debug)]
+struct MockAntenna {
+    hard_coded_status: u16,
+    hard_coded_body: BagOfBytes,
+    spy: fn(NetworkRequest) -> (),
+}
+
+#[allow(unused)]
+impl MockAntenna {
+    fn with_spy(
```
Comment (CyonAlexRDX) [inline src/gateway_api/client/gateway_client.rs:65]: this is new, unit tests below can "spy" on the Request

File: src/gateway_api/client/gateway_client.rs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            hard_coded_body: body.into(),
+            spy,
+        }
+    }
+
+    fn new(status: u16, body: impl Into<BagOfBytes>) -> Self {
+        Self::with_spy(status, body, |_| {})
+    }
+
+    fn with_response<T>(response: T) -> Self
```
Comment (CyonAlexRDX) [inline src/gateway_api/client/gateway_client.rs:81]: new: the unit tests below can specify (mock) the response

File: src/gateway_api/client/gateway_client.rs [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+    #[actix_rt::test]
+    async fn test_submit_notarized_transaction_mock_duplicate() {
+        let mock_antenna =
+            MockAntenna::with_response(TransactionSubmitResponse {
+                duplicate: true,
+            });
+        let sut = SUT::new(Arc::new(mock_antenna), NetworkID::Stokenet);
+        let req =
+            sut.submit_notarized_transaction(NotarizedTransaction::sample());
```
Comment (CyonAlexRDX) [inline src/gateway_api/client/gateway_client.rs:212]: testing new method, by mocking that it is a duplicate and assert we get `CommonError::GatewaySubmitDuplicateTX`

File: src/gateway_api/client/gateway_client_dispatch_request.rs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        serde_json::from_slice::<U>(&response.body).map_err(|_| {
+            CommonError::NetworkResponseJSONDeserialize {
+                into_type: std::any::type_name::<U>().to_owned(),
+            }
+        })
+    }
+
+    async fn dispatch_network_request<T, U, V, F>(
```
Comment (CyonAlexRDX) [inline src/gateway_api/client/gateway_client_dispatch_request.rs:73]: renamed from `make_request` to "dispatch", which I think is a good name... tells FFI side (iOS/Android) to execute the request.

File: src/gateway_api/endpoints/state_endpoints.rs
Comment (CyonAlexRDX) [inline src/gateway_api/endpoints/state_endpoints.rs:1]: split out each endpoint into a by top level path grouped folder structure, e.g. here is `state/*` endpoints

File: src/gateway_api/endpoints/transaction_endpoints.rs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    ///
+    /// Returns information needed to construct a new transaction including current epoch number.
+    ///
+    /// See [the Gateway API docs for details][doc].
+    ///
+    /// [doc]: https://radix-babylon-gateway-api.redoc.ly/#operation/TransactionConstruction
+    pub(crate) async fn transaction_construction(&self) -> Result<LedgerState> {
+        self.post_empty(
+            "transaction/construction",
+            |response: TransactionConstructionResponse| {
```
Comment (CyonAlexRDX) [inline src/gateway_api/endpoints/transaction_endpoints.rs:14]: and in `src/gateway_api/endpoints/transaction_endpoints` all `transaction/` endpoints go...

File: src/gateway_api/endpoints/transaction_endpoints.rs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /// Previews transaction against the network - aka "dry run" of transaction.
+    ///
+    /// Previews are used for [`execution_summary`][TransactionManifest::execution_summary], which the iOS/Android wallet app
+    /// uses to present the "review transaction" screen to the user.
+    ///
+    /// See [the Gateway API docs for details][doc].
+    ///
+    /// [doc]: https://radix-babylon-gateway-api.redoc.ly/#operation/TransactionPreview
```
Comment (CyonAlexRDX) [inline src/gateway_api/endpoints/transaction_endpoints.rs:28]: also added doc to all methods

File: src/gateway_api/methods/state_methods.rs [hunk 13]
```diff
@@ -0,0 +1,56 @@
+use crate::prelude::*;
+
+#[uniffi::export]
+impl GatewayClient {
```
Comment (CyonAlexRDX) [inline src/gateway_api/methods/state_methods.rs:4]: async func which *uses* endpoints, but add some custom logic "on top of them" go into the mod `gateway_api/methods`

Comment (CyonAlexRDX) [reply src/gateway_api/methods/state_methods.rs:4]: in this file, method using `/state` endpoint...

File: src/gateway_api/models/logic/request/gw_public_key.rs [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+impl From<PublicKey> for GWPublicKey {
+    fn from(value: PublicKey) -> Self {
+        match value {
+            PublicKey::Ed25519(key) => Self::Ed25519(key),
+            PublicKey::Secp256k1(key) => Self::Secp256k1(key),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for GWPublicKey {
```
Comment (CyonAlexRDX) [inline src/gateway_api/models/logic/request/gw_public_key.rs:12]: manual impl of `serde` allows us to skip a few unnecessary models...

File: src/gateway_api/models/logic/request/state/entity/state_entity_details.rs
Comment (CyonAlexRDX) [inline src/gateway_api/models/logic/request/state/entity/state_entity_details.rs:1]: now follows a bunch of JSON unit tests for requests and responses.

File: src/gateway_api/network_antenna/network_antenna.rs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+/// (iOS/Android), so that Sargon can with some HTTP client perform make network
+/// requests.
+#[uniffi::export(with_foreign)]
+#[async_trait::async_trait]
+pub trait NetworkAntenna: Send + Sync {
+    /// Method invoked by Sargon Rust side, telling an implementing type to
+    /// execute a `NetworkRequest` and pass the Result back to Sargon Rust side.
+    ///
+    /// Either: `Err` or `Ok(NetworkResponse)`.
+    async fn execute_network_request(
```
Comment (CyonAlexRDX) [inline src/gateway_api/network_antenna/network_antenna.rs:13]: antenna moved to seperate file, documented and renamed method -> `execute_network_request`

File: tests/integration/main.rs [hunk 16]
```diff
@@ -127,4 +127,87 @@ mod integration_tests {
             )])
         );
     }
+
+    async fn submit_tx_use_faucet(
```
Comment (CyonAlexRDX) [inline tests/integration/main.rs:131]: usage of faucet! which works! 🥳

**Review by micbakos-rdx — APPROVED** [review]

**Pull request merged**